chapters/unitexpressions.tex

```diff
@@ -89,3 +89,171 @@ \section{The Syntax of Unit Expressions}\label{the-syntax-of-unit-expressions}
+\item
+  The result of multiplication has a unit obtained by multiplying the operands' units.
+\item
+  Built-in operators such as \lstinline!pre! and \lstinline!previous! preserve units, while \lstinline!der! changes the unit by dividing it by \lstinline!"s"!.
+\end{itemize}
+
+
+\subsection{Empty and Undefined Units}\label{empty-and-undefined-units}
+
+In situations where the specification does not prescribe how to determine the unit of an expression, the unit of that expression is said to be \firstuse[undefined unit]{undefined}.
+It is then not possible for a tool to reject the equation (or other construct) with support in the specification, and options for the tool include silently not performing unit checking, or applying checks based dimensional analysis.
+
+The unit of an expression can also be defined as being \firstuse[empty unit]{empty}, see below.
+In certain places (see below), expressions with empty unit can be implicitly cast to suitable units.
+When an expression with empty unit is implicitly cast to some unit, that unit is referred to as the \firstuse{inferred unit} of the expression.
+
+
+\subsection{Unit Inference}\label{unit-inference}
+
+Where the empty unit has no other defined meaning, it can be implicitly cast to unit \lstinline!"1"!.
+
+In some cases the empty unit can be implicitly cast to an inferred unit:
+\begin{itemize}
+\item
+  In binding equations and modifications:
+  \begin{itemize}
+  \item The entire expression of the binding or modification.
+  \item When the entire expression is an array construction, array concatenation and array range, then apply rules recursively for the direct subexpressions.
+  \end{itemize}
+\item
+  The entire argument expression in a function call.
+% To keep the design close to the bare minimum, this part is currently excluded:
+%\item
+%  For a translation-time constant with value 0.0:
+%  \begin{itemize}
+%  \item Expressions constituting the entire side of an equality or relation.
+%  \item The right hand side of an assignment.
+%  \item The direct subexpressions of array construction, array concatenation, and array range.
+%  \end{itemize}
+\end{itemize}
+
+
+\subsection{Expressions with Empty Unit}\label{expressions-with-empty-unit}
+
+This section describes conditions under which an expression has empty unit.
+Conditions not given here must not be interpreted as definitely not implying empty unit; instead, the unit may be currently undefined for some expressions, allowing the unit to be properly defined in future versions of the specification.
+
+Basic expressions defined to have empty unit:
+\begin{itemize}
+\item
+  \lstinline!Real! literals.
+\item
+  \lstinline!Integer! expressions implicitly cast to \lstinline!Real!.
+\end{itemize}
+
+Expressions defined to \emph{not} propagate the empty unit, thereby forcing inference of unit \lstinline!"1"!:
+\begin{itemize}
+\item
+  Addition, subtraction, multiplication and division operators when either operand has non-empty unit.
```

Comment:
Actually, in these cases you don't 'force inference of unit "1"', nor propagate the empty unit, but these expressions (i.e. the operators + operands) propagate the unit of the operand with the non-empty unit (addition, subtraction, multiplication, division when in numerator) or its reciprocal (division when in denominator).

Reply:
The way it is formulated comes from the need to leave gray areas open for future design, with the understanding that tools may take the easy way out and propagate empty units up the expression three whenever the specification doesn't define what to do, in order to not reject things without support in the specification.  I am thus categorizing all expressions (where one or more subexpressions have empty unit) according to:
- Expressions where the empty unit *is* propagated up the expression tree.
- Expressions where the empty unit *is not* propagated up the expression tree.
- Expressions not yet covered by the specification (meaning their unit is so far undefined).

Maybe there are alternative ways of achieving the same effect, but I'm not sure how to do it without ending up with a proposal that covers more than what the language group wants as a solution for MCP-0027.  In the future, when all cases are covered by the specification, there should be an easier way to express the rule:
- Describe situations where the empty unit gets propagated up the expression tree.
- In all other situations, unit `"1"` is inferred.

Reply:
To be clear, it's fine to have grey areas (if you mean areas that the spec does not cover), they just should be clearly delineated. IMO, having grey areas is not a contradiction to having a clearly defined process for the rest.

I think I understand what you mean better, but I'll have to think on this a bit, I'm a bit confused.

Reply:
This that mean we are ready to close this conversation as resolved?